gr_modtool generates cpp noblock that does not compile

### What happened?

I am trying to generate a noblock using gr_modtool.

```
gr_modtool newmod my_rfid

gr_modtool add testblock
GNU Radio module name identified: my_rfid
('sink', 'source', 'sync', 'decimator', 'interpolator', 'general', 'tagged_stream', 'hier', 'noblock')
Enter block type: noblock
Language (python/cpp): cpp
Language: C++
Block/code identifier: testblock
Please specify the copyright holder: none
Enter valid argument list, including default arguments: 
 
Add C++ QA code? [Y/n] n
Adding file 'lib/testblock.cc'...
Adding file 'include/gnuradio/my_rfid/testblock.h'...
Adding file 'python/my_rfid/bindings/docstrings/testblock_pydoc_template.h'...
Adding file 'python/my_rfid/bindings/testblock_python.cc'...
Adding file 'grc/my_rfid_testblock.block.yml'...
Editing grc/CMakeLists.txt...
```

When I attempt a build:

```
[ 70%] Built target my_rfid_docstrings
[ 80%] Building CXX object python/my_rfid/bindings/CMakeFiles/my_rfid_python.dir/testblock_python.cc.o
/home/mitsos/Documents/Diplomatiki/gr-my_rfid/python/my_rfid/bindings/testblock_python.cc: In function ‘void bind_testblock(pybind11::module&)’:
/home/mitsos/Documents/Diplomatiki/gr-my_rfid/python/my_rfid/bindings/testblock_python.cc:37:34: error: template argument 2 is invalid
   37 |         std::shared_ptr<testblock>>(m, "testblock", D(testblock))
      |                                  ^~
/home/mitsos/Documents/Diplomatiki/gr-my_rfid/python/my_rfid/bindings/testblock_python.cc:39:35: error: ‘make’ is not a member of ‘testblock’ {aka ‘gr::my_rfid::testblock’}
   39 |         .def(py::init(&testblock::make),
      |                                   ^~~~
make[2]: *** [python/my_rfid/bindings/CMakeFiles/my_rfid_python.dir/build.make:76: python/my_rfid/bindings/CMakeFiles/my_rfid_python.dir/testblock_python.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:384: python/my_rfid/bindings/CMakeFiles/my_rfid_python.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

### System Information

OS: Arch Linux x86_64
GR Installation Method: AUR: https://aur.archlinux.org/packages/gnuradio-git


### GNU Radio Version

3.11-git (main)

### Specific Version

v3.11.0.0git-576-g9fca5db4

### Steps to Reproduce the Problem

Generate a new gr module and add a cpp noblock. Then try to compile.

### Relevant log output

_No response_

--- a/gr-utils/bindtool/templates/pydoc_h.mako
+++ b/gr-utils/bindtool/templates/pydoc_h.mako
@@ -22,14 +22,14 @@ ${license}
  */
 ${render_namespace(namespace=namespace, modname=[modname])} \
 \
-<%def name='render_docstring_const(modname,names,info,docstring="",info_all=None)'>
+<%def name='render_docstring_const(modname,names,info,docstring="",info_all=None,max_unambiguous=1)'>
 <% 
   names = list(filter(None, names))
   suffix = ''
   if info_all:
     matcher = lambda x,name: x['name'] == name
     matched_list = [f for f in info_all if matcher(f,info['name'])]
-    overloaded = len(matched_list) > 1
+    overloaded = len(matched_list) > max_unambiguous
     suffix = ''
     if overloaded:
       index_into_list = matched_list.index(info)
@@ -56,7 +56,7 @@ static const char *__doc_${'_'.join(names)}${suffix} = R"doc(${docstring})doc";
 ${render_docstring_const(modname,namespace['name'].split('::')+[cls['name']],cls,cls['docstring'] if 'docstring' in cls else "")}
 \
 % for cotr in constructors:
-${render_docstring_const(modname,namespace['name'].split('::')+[cls['name'],cotr['name']],cotr,cotr['docstring'] if 'docstring' in cotr else "",constructors)}
+${render_docstring_const(modname,namespace['name'].split('::')+[cls['name'],cotr['name']],cotr,cotr['docstring'] if 'docstring' in cotr else "",constructors,0)}
 % endfor ## constructors
 \
 % for fcn in member_functions:

--- a/gr-utils/modtool/core/add.py
+++ b/gr-utils/modtool/core/add.py
@@ -10,7 +10,6 @@
 
 
 import os
-import re
 import logging
 import subprocess
 
@@ -29,11 +28,11 @@ def clang_format(s):
             ["clang-format"], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
         out, err = p.communicate(s.encode('utf-8'))
         if p.returncode != 0:
-            print("Failed to run clang-format: %s", err)
+            print(f"Failed to run clang-format: {err}")
             return s
         return out.decode('utf-8')
     except (RuntimeError, FileNotFoundError) as e:
-        print("Failed to run clang-format: %s", e)
+        print(f"Failed to run clang-format: {e}")
         return s
 
 
@@ -128,11 +127,12 @@ class ModToolAdd(ModTool):
         """ Shorthand for writing a substituted template to a file"""
         path_to_file = os.path.join(path, fname)
         logger.info(f"Adding file '{path_to_file}'...")
-        def formatter(x): return x
-        if fname.endswith('.cc') or fname.endswith('.h'):
-            formatter = clang_format
         with open(path_to_file, 'w') as f:
-            f.write(formatter(render_template(tpl, **self.info)))
+            if fname.endswith('.cc') or fname.endswith('.h'):
+                f.write(clang_format(render_template(tpl, **self.info)))
+            else:
+                f.write(render_template(tpl, **self.info))
+
         self.scm.add_files((path_to_file,))
 
     def run(self):
@@ -211,7 +211,7 @@ class ModToolAdd(ModTool):
         """
         fname_cc = None
         fname_h = None
-        if self.info['version'] in ('37', '38', '310'):
+        if self.info['version'] in ('38', '310'):
             fname_h = self.info['blockname'] + '.h'
             fname_cc = self.info['blockname'] + '.cc'
             if self.info['blocktype'] in ('source', 'sink', 'sync', 'decimator',
@@ -221,22 +221,10 @@ class ModToolAdd(ModTool):
                                 self.info['blockname'] + '_impl.h')
             self._write_tpl('block_impl_cpp', 'lib', fname_cc)
             self._write_tpl('block_def_h', self.info['includedir'], fname_h)
-        else:  # Pre-3.7 or autotools
-            fname_h = self.info['fullblockname'] + '.h'
-            fname_cc = self.info['fullblockname'] + '.cc'
-            self._write_tpl('block_h36', self.info['includedir'], fname_h)
-            self._write_tpl('block_cpp36', 'lib', fname_cc)
+        else:  # Pre-3.8 or autotools
+            raise RuntimeError(f"Unsupported version {self.info['version']}")
         if self.add_cc_qa:
-            if self.info['version'] in ['38', '310']:
-                self._run_cc_qa_boostutf()
-            elif self.info['version'] == '37':
-                self._run_cc_qa()
-            elif self.info['version'] == '36':
-                logger.warning(
-                    "Warning: C++ QA files not supported for 3.6-style OOTs.")
-            elif self.info['version'] == 'autofoo':
-                logger.warning(
-                    "Warning: C++ QA files not supported for autotools.")
+            self._run_cc_qa_boostutf()
         if not self.skip_cmakefiles:
             ed = CMakeFileEditor(self._file['cmlib'])
             cmake_list_var = '[a-z]*_?' + self.info['modname'] + '_sources'
@@ -256,6 +244,7 @@ class ModToolAdd(ModTool):
         - add reference and call to bind_blockname()
         - include them into CMakeLists.txt
         """
+        import hashlib
 
         bindings_dir = os.path.join(self.info['pydir'], 'bindings')
 
@@ -285,7 +274,6 @@ class ModToolAdd(ModTool):
                               flag_automatic=(blktype != 'noblock'))
         block_base = code_generator.GRTYPELIST.get(blktype, '')
 
-        import hashlib
         header_file = self.info['blockname'] + '.h'
         hasher = hashlib.md5()
         with open(os.path.join(self.info['includedir'], header_file), 'rb') as file_in:

--- a/gr-utils/modtool/core/add.py
+++ b/gr-utils/modtool/core/add.py
@@ -278,13 +278,12 @@ class ModToolAdd(ModTool):
             prefix_include_root = '/'.join(('gnuradio', self.info['modname']))
         else:
             prefix_include_root = self.info['modname']
+        blktype = self.info['blocktype']
 
         bg = BindingGenerator(prefix=gr.prefix(), namespace=[
-                              'gr', self.info['modname']], prefix_include_root=prefix_include_root)
-        block_base = ""
-        if self.info['blocktype'] in ('source', 'sink', 'sync', 'decimator',
-                                      'interpolator', 'general', 'hier', 'tagged_stream'):
-            block_base = code_generator.GRTYPELIST[self.info['blocktype']]
+                              'gr', self.info['modname']], prefix_include_root=prefix_include_root,
+                              flag_automatic=(blktype != 'noblock'))
+        block_base = code_generator.GRTYPELIST.get(blktype, '')
 
         import hashlib
         header_file = self.info['blockname'] + '.h'
@@ -305,19 +304,6 @@ class ModToolAdd(ModTool):
                 "classes": [
                     {
                         "name": self.info['blockname'],
-                        "member_functions": [
-                            {
-                                "name": "make",
-                                "return_type": "::".join(("gr", self.info['modname'], self.info['blockname'], "sptr")),
-                                "has_static": "1",
-                                "arguments": []
-                            }
-                        ],
-                        "bases": [
-                            "::",
-                            "gr",
-                            block_base
-                        ],
                         "constructors": [
                             {
                                 "name": self.info['blockname'],
@@ -330,6 +316,21 @@ class ModToolAdd(ModTool):
                 "namespaces": []
             }
         }
+
+        if block_base:
+            header_info['namespace']['classes'][0]['bases'] = ["::", "gr", block_base]
+
+        if blktype != 'noblock':
+            # Only blocks have make
+            header_info['namespace']['classes'][0]['member_functions'] = [
+                {
+                    "name": "make",
+                    "return_type": f"gr::{self.info['modname']}::{self.info['blockname']}::sptr",
+                    "has_static": "1",
+                    "arguments": []
+                }
+            ]
+
         # def gen_pybind_cc(self, header_info, base_name):
         pydoc_txt = bg.gen_pydoc_h(header_info, self.info['blockname'])
         path_to_file = os.path.join(bindings_dir, fname_pydoc_h)

